crates/oxc_linter/src/rules/unicorn/no_useless_undefined.rs

```diff
@@ -0,0 +1,403 @@
+        (r"const foo = () => {};", None),
+        (r"let foo;", None),
+        (r"var foo;", None),
+        (r"const foo = undefined;", None),
+        (r"foo();", None),
+        (r"foo(bar,);", None),
+        (r"foo(undefined, bar);", None),
+        (r"const {foo} = {};", None),
+        (r"function foo({bar} = {}) {}", None),
+        (r"function foo(bar) {}", None),
+        // I guess nobody uses this, but `yield* undefined;` is valid code, and `yield*;` is not
+        (r"function* foo() {yield* undefined;}", None),
+        // Ignored
+        (r"if (Object.is(foo, undefined)){}", None),
+        (r"t.is(foo, undefined)", None),
+        (r"assert.equal(foo, undefined, message)", None),
+        (r"assert.notEqual(foo, undefined, message)", None),
+        (r"assert.strictEqual(foo, undefined, message)", None),
+        (r"assert.notStrictEqual(foo, undefined, message)", None),
+        (r"assert.propertyVal(foo, 'bar', undefined, message)", None),
+        (r"assert.notPropertyVal(foo, 'bar', undefined, message)", None),
+        (r"expect(foo).not(undefined)", None),
+        (r"expect(foo).to.have.property('bar', undefined)", None),
+        (r"expect(foo).toBe(undefined)", None),
+        (r"expect(foo).toContain(undefined)", None),
+        (r"expect(foo).toContainEqual(undefined)", None),
+        (r"expect(foo).toEqual(undefined)", None),
+        (r"t.same(foo, undefined)", None),
+        (r"t.notSame(foo, undefined)", None),
+        (r"t.strictSame(foo, undefined)", None),
+        (r"t.strictNotSame(foo, undefined)", None),
+        (r"expect(someFunction).toHaveBeenCalledWith(1, 2, undefined);", None),
+        (r"set.add(undefined);", None),
+        (r"map.set(foo, undefined);", None),
+        (r"array.push(foo, undefined);", None),
+        (r"array.push(undefined);", None),
+        (r"array.unshift(foo, undefined);", None),
+        (r"array.unshift(undefined);", None),
+        (r"createContext(undefined);", None),
+        (r"React.createContext(undefined);", None),
+        (r"setState(undefined)", None),
+        (r"setState?.(undefined)", None),
+        (r"props.setState(undefined)", None),
+        (r"props.setState?.(undefined)", None),
+        (r"array.includes(undefined)", None),
+        (r"set.has(undefined)", None),
+        // `Function#bind()`
+        (r"foo.bind(bar, undefined);", None),
+        (r"foo.bind(...bar, undefined);", None),
+        (r"foo.bind(...[], undefined);", None),
+        (r"foo.bind(...[undefined], undefined);", None),
+        (r"foo.bind(bar, baz, undefined);", None),
+        (r"foo?.bind(bar, undefined);", None),
+        // `checkArguments: false`
+        (r"foo(undefined, undefined);", options_ignore_arguments()),
+        (r"foo.bind(undefined);", options_ignore_arguments()),
+        // `checkArrowFunctionBody: false`
+        (r"const foo = () => undefined", options_ignore_arrow_function_body()),
+    ];
+
+    let fail = vec![(r"let foo = undefined;", None)];
```

Comment:
Should we be checking for cases such as parameter initializers and assignment expressions?

Reply:
Yes it has been added.